File: app/students/students_service.py
from uuid import uuid4

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.shared.exceptions import conflict
from app.students.student_model import Student as StudentModel
from app.students.students_schemas import (
    CreateStudentDto,
    Student,
    StudentList,
    UpdateStudentDto,
)


class StudentsService:
    def find_all(self, db: Session) -> StudentList:
        students = db.scalars(
            select(StudentModel).order_by(StudentModel.createdAt.desc())
        ).all()

        return StudentList(
            total=len(students),
            items=[Student.model_validate(s) for s in students],
        )

    def find_by_id(self, db: Session, student_id: str) -> Student:
        return Student.model_validate(self._get(db, student_id))

    def create(self, db: Session, data: CreateStudentDto) -> Student:
        self.assert_no_conflicts(db, data)

        student = StudentModel(
            id=str(uuid4()),
            name=data.name,
            username=data.username,
            email=data.email,
            age=data.age,
        )
        db.add(student)
        db.commit()

        return Student.model_validate(student)

    def update(self, db: Session, student_id: str, data: UpdateStudentDto) -> Student:
        student = self._get(db, student_id)
        self.assert_no_conflicts(db, data, except_id=student_id)

        if data.name is not None:
            student.name = data.name
        if data.username is not None:
            student.username = data.username
        if data.email is not None:
            student.email = data.email
        if data.age is not None:
            student.age = data.age

        db.commit()

        return Student.model_validate(student)

    def delete(self, db: Session, student_id: str) -> Student:
        student = self._get(db, student_id)
        result = Student.model_validate(student)

        db.delete(student)
        db.commit()

        return result

    def assert_no_conflicts(
        self,
        db: Session,
        data: CreateStudentDto | UpdateStudentDto,
        except_id: str | None = None,
    ) -> None:
        errors: dict[str, dict[str, str]] = {}

        if data.email:
            existing = db.scalar(
                select(StudentModel).where(StudentModel.email == data.email)
            )
            if existing and existing.id != except_id:
                errors["email"] = {"message": "El correo electrónico ya está en uso"}

        if data.username:
            existing = db.scalar(
                select(StudentModel).where(StudentModel.username == data.username)
            )
            if existing and existing.id != except_id:
                errors["username"] = {"message": "El nombre de usuario ya está en uso"}

        if errors:
            raise conflict(errors)

    def _get(self, db: Session, student_id: str) -> StudentModel:
        student = db.get(StudentModel, student_id)

        if student is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Estudiante no encontrado",
            )

        return student


students_service = StudentsService()

File: src/students/students.service.ts
import { randomUUID } from "node:crypto"
import { ConflictException, Injectable, NotFoundException } from "@nestjs/common"
import { InjectRepository } from "@nestjs/typeorm"
import { Repository } from "typeorm"

import { Student } from "./students.entity"
import { CreateStudentDto, UpdateStudentDto } from "@/students/students.dtos"
import type { ValidationErrors } from "@/shared/response"

@Injectable()
export class StudentsService {
	constructor(
		@InjectRepository(Student)
		private readonly studentsRepository: Repository<Student>,
	) {}

	public findAll(): Promise<Student[]> {
		return this.studentsRepository.find({ order: { createdAt: "DESC" } })
	}

	public async findById(id: string): Promise<Student> {
		const student = await this.studentsRepository.findOne({ where: { id } })

		if (!student) {
			throw new NotFoundException("Estudiante no encontrado")
		}

		return student
	}

	public async create(data: CreateStudentDto): Promise<Student> {
		await this.assertNoConflicts(data)

		const student = this.studentsRepository.create({
			id: randomUUID(),
			name: data.name,
			username: data.username,
			email: data.email,
			age: data.age,
		})

		return this.studentsRepository.save(student)
	}

	public async update(id: string, data: UpdateStudentDto): Promise<Student> {
		const existing = await this.findById(id)
		await this.assertNoConflicts(data, id)

		Object.assign(existing, {
			name: data.name ?? existing.name,
			username: data.username ?? existing.username,
			email: data.email ?? existing.email,
			age: data.age ?? existing.age,
		})

		return this.studentsRepository.save(existing)
	}

	public async delete(id: string): Promise<Student> {
		const existing = await this.findById(id)
		await this.studentsRepository.delete(id)
		return existing
	}

	private async assertNoConflicts(
		data: { email?: string; username?: string },
		exceptId?: string,
	): Promise<void> {
		const conflicts: ValidationErrors = {}

		if (data.email) {
			const existing = await this.studentsRepository.findOne({ where: { email: data.email } })

			if (existing && existing.id !== exceptId) {
				conflicts.email = { message: "El correo electrónico ya está en uso" }
			}
		}

		if (data.username) {
			const existing = await this.studentsRepository.findOne({
				where: { username: data.username },
			})

			if (existing && existing.id !== exceptId) {
				conflicts.username = { message: "El nombre de usuario ya está en uso" }
			}
		}

		if (Object.keys(conflicts).length > 0) {
			throw new ConflictException({
				message: "Conflicto con datos existentes",
				errors: conflicts,
			})
		}
	}
}
